managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -1121,11 +1184,49 @@ public boolean cancelPendingReadRequest() {
         if (log.isDebugEnabled()) {
             log.debug("[{}] [{}] Cancel pending read request", ledger.getName(), name);
         }
-        final OpReadEntry op = WAITING_READ_OP_UPDATER.getAndSet(this, null);
-        if (op != null) {
-            op.recycle();
+        synchronized (pendingReadOpMutex) {
+            final OpReadEntry op = WAITING_READ_OP_UPDATER.get(this);
+            // Case 1: the pending read has executed, or there is no pending read.
+            if (op == null) {
+                return false;
+            }
+            Function<Boolean, Boolean> clearWaitingReadOp = removeWaitingCursor -> {
+                if (WAITING_READ_OP_UPDATER.compareAndSet(this, op,null)) {
```

Comment:
```suggestion
                if (WAITING_READ_OP_UPDATER.compareAndSet(this, op, null)) {
```

I think checkstyle should fail. Have you run the checkstyle check locally?

Reply:
Addressed